plugins/Kaleidoscope-MouseKeys/src/kaleidoscope/plugin/MouseKeys.cpp

```diff
@@ -183,50 +124,85 @@ EventHandlerResult MouseKeys_::onKeyEvent(KeyEvent &event) {
   if (!isMouseKey(event.key))
     return EventHandlerResult::OK;
 
+  pending_directions_ = 0;
+
   if (isMouseButtonKey(event.key)) {
-    sendMouseButtonReport(event);
+    // Clear button state; it will be repopulated by `onAddToReport()`, and the
+    // report will be sent by `afterReportingState()`.
+    buttons_ = 0;
 
   } else if (isMouseWarpKey(event.key)) {
     if (keyToggledOn(event.state)) {
       sendMouseWarpReport(event);
     }
+  }
+
+  // Reports for mouse cursor and wheel movement keys are sent from the
+  // `afterReportingState()` handler (when first toggled on) and
+  // `afterEachCycle()` handler (when held).  We need to return `OK` here so
+  // that processing of events for these keys will complete.
+  return EventHandlerResult::OK;
+}
+
+// -----------------------------------------------------------------------------
+EventHandlerResult MouseKeys_::afterReportingState(const KeyEvent &event) {
```

Comment:
In an evented world, "afterReportingState" does seem a more right time to act than "onKeyEvent", but it makes me feel like we should be doing the setup in "beforeReportingState" and then having the core actually send Mouse reports like we do for Keyboard, SystemControl and ConsumerControl.

To be clear: I don't think we want to make that heavy a change a precondition of getting this merged. 

I'd love to hear your thinking about why this is afterReportingState, rather than beforeReportingState or if it's essentially a no-op and you just picked one arbitrarily.

Reply:
I'm pretty sure it could just as easily work in `beforeReportingState()`, but there is one advantage (or perhaps disadvantage) to doing it in `afterReportingState()` instead: the former is abortable (if a plugin returns `ABORT`, the others won't get called).

I think there's a reason we don't want the mouse reports sent in the core: if you don't have MouseKeys in your sketch, it might be able to optimize all the mouse stuff out, making the binary much smaller for people who have no use for mouse keys.  I haven't checked to be sure that actually happens, though.

I'm okay with changing it to `beforeReportingState()`, if you prefer it that way, as long as it works (I haven't tested it to confirm).

Reply:
I don't have a strong feeling there. I wanted to understand the rationale, though. - It's probably worth sticking that in a comment along with the code?

Indeed, I agree that we ought to be optimizing mouse stuff out if it's not used.

Reply:
One more noteworthy thing: mouse movement and scroll reports get sent at times other than in response to a key toggling on or off.